### PR TITLE
Added soft_nearest_neighbor_loss. Closes #103

### LINKED ISSUE #103: Implement Soft Nearest Neighbor Loss
Analyzing and Improving Representations with the Soft Nearest Neighbor Loss https://arxiv.org/pdf/1902.01889.pdf

### PR DESCRIPTION
**Pull request by abhisharsinha**
Implemented soft nearest neighbor loss as defined in [Analyzing and Improving Representations with the Soft Nearest Neighbor Loss](https://arxiv.org/pdf/1902.01889.pdf) Closes issue #103 

Added test in test_losses.py for testing which can be run as pytest tests/test_losses.py

### CODE REVIEW

File: tensorflow_similarity/losses/softnn_loss.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+import tensorflow as tf
+from tensorflow.keras.losses import Loss
+from tensorflow_similarity.distances import Distance, distance_canonicalizer
+
+from tensorflow_similarity.types import FloatTensor, IntTensor
+
+
+@tf.keras.utils.register_keras_serializable(package="Similarity")
+class SoftNearestNeighborLoss(Loss):
```
Comment (owenvallis) [inline tensorflow_similarity/losses/softnn_loss.py:11]: Let's make this a subclass of MetricLoss (see multisim_loss.py). We're going to work on a small refactor of the MetricLoss base class to better support multi GPU training and it would good to have all of these follow the same format.

File: tests/test_losses.py [hunk 2]
```diff
@@ -124,3 +125,50 @@ def test_np_loss():
     # y_true, y_preds
     loss = pnl(y_true, y_preds)
     assert loss > 0.9
+
+
+# [soft neasrest neighbor loss]
+def test_softnn_loss_serialization():
+    loss = SoftNearestNeighborLoss()
+    config = loss.get_config()
+    print(config)
```
Comment (owenvallis) [inline tests/test_losses.py:134]: Stray print in the test.

File: tests/test_losses.py [hunk 3]
```diff
@@ -124,3 +125,50 @@ def test_np_loss():
...
+    loss2 = SoftNearestNeighborLoss.from_config(config)
+    assert loss.name == loss2.name
+    assert loss.distance == loss2.distance
```
Comment (owenvallis) [inline tests/test_losses.py:137]: I think this is only loading the distance from default kwarg value and not from the config. I'm not sure if this will pass if we set the distance on line 132 to something else like 'cosine'. 

I think subclassing from MetricLoss should fix this as we add all kwargs to the config.

File: tests/test_losses.py [hunk 4]
```diff
@@ -124,3 +125,50 @@ def test_np_loss():
...
+def test_softnn_loss():
+    num_inputs = 10
+    # y_true: labels
+    y_true = tf.random.uniform((num_inputs, ), 0, 10, dtype=tf.int32)
+    # y_preds: embedding
+    y_preds = tf.random.uniform((num_inputs, 20), 0, 1)
+    temperature = np.random.uniform(0.5,5)
+    softnn = SoftNearestNeighborLoss()
+    loss = softnn(y_true, y_preds, temperature)
+    assert loss.numpy() - softnn_util(y_true, y_preds, temperature) < 1e-9
```
Comment (owenvallis) [inline tests/test_losses.py:174]: [nit] looks like github want's a newline at the end of the file.

**Review by owenvallis — CHANGES_REQUESTED** [review]
Thanks for the PR! Added some comments about subclassing from MetricLoss. This is both because we will be making some updates soon and also so that we ensure we store the distance in the config. I think it will also auto add your temp param as well.

**Comment by abhisharsinha** [comment]
@owenvallis Thanks for reviewing. I have made the suggested changes.

**Review by owenvallis — APPROVED** [review]
Thanks again for adding the new loss.
